x-pack/plugins/actions/server/action_service.ts

```diff
@@ -0,0 +1,154 @@
+  references?: SavedObjectReference[];
+}
+
+export class ActionService {
+  private actionTypeService: ActionTypeService;
+  private encryptedSavedObjectsPlugin: EncryptedSavedObjectsPlugin;
+
+  constructor(
+    actionTypeService: ActionTypeService,
+    encryptedSavedObjectsPlugin: EncryptedSavedObjectsPlugin
+  ) {
+    this.actionTypeService = actionTypeService;
+    this.encryptedSavedObjectsPlugin = encryptedSavedObjectsPlugin;
+  }
+
+  /**
+   * Create an action
+   */
+  public async create(
+    savedObjectsClient: SavedObjectsClient,
+    data: Action,
+    options?: CreateOptions
+  ) {
+    const { actionTypeId } = data;
+    if (!this.actionTypeService.has(actionTypeId)) {
+      throw Boom.badRequest(`Action type "${actionTypeId}" is not registered.`);
+    }
+    this.actionTypeService.validateActionTypeConfig(actionTypeId, data.actionTypeConfig);
+    const actionWithSplitActionTypeConfig = this.moveEncryptedAttributesToSecrets(data);
+    return await savedObjectsClient.create('action', actionWithSplitActionTypeConfig, options);
+  }
+
+  /**
+   * Get an action
+   */
+  public async get(savedObjectsClient: SavedObjectsClient, id: string) {
+    return await savedObjectsClient.get('action', id);
+  }
+
+  /**
+   * Find actions
+   */
+  public async find(savedObjectsClient: SavedObjectsClient, options: FindOptions = {}) {
+    return await savedObjectsClient.find({
+      ...options,
+      type: 'action',
+    });
+  }
+
+  /**
+   * Delete action
+   */
+  public async delete(savedObjectsClient: SavedObjectsClient, id: string) {
+    return await savedObjectsClient.delete('action', id);
+  }
+
+  /**
+   * Update action
+   */
+  public async update(
+    savedObjectsClient: SavedObjectsClient,
+    id: string,
+    data: Action,
+    options: { version?: string; references?: SavedObjectReference[] } = {}
+  ) {
+    const { actionTypeId } = data;
+    if (!this.actionTypeService.has(actionTypeId)) {
+      throw Boom.badRequest(`Action type "${actionTypeId}" is not registered.`);
+    }
+    this.actionTypeService.validateActionTypeConfig(actionTypeId, data.actionTypeConfig);
+    const actionWithSplitActionTypeConfig = this.moveEncryptedAttributesToSecrets(data);
+    return await savedObjectsClient.update('action', id, actionWithSplitActionTypeConfig, options);
+  }
+
+  /**
+   * Fire an action
+   */
+  public async fire({ id, params, savedObjectsClient }: FireOptions) {
+    const action = await this.encryptedSavedObjectsPlugin.getDecryptedAsInternalUser('action', id);
+    const mergedActionTypeConfig = {
+      ...action.attributes.actionTypeConfig,
+      ...action.attributes.actionTypeConfigSecrets,
+    };
+    return await this.actionTypeService.execute({
+      id: action.attributes.actionTypeId,
+      actionTypeConfig: mergedActionTypeConfig,
+      params,
+    });
+  }
+
+  /**
+   * Set actionTypeConfigSecrets values on a given action
+   */
+  private moveEncryptedAttributesToSecrets(action: Action) {
```

Comment:
Function should be removed and let the API have the two attributes. Expose the unencryptedAttributes within the list action types request which will eventually expose the params each action type expects. This will also solve the problem when saved objects API is called directly and attributes **don't** get encrypted.